src/core/initialize.rs

```diff
@@ -74,7 +83,25 @@ impl StackTraceGetter {
         Ok(self.get_trace_from_current_thread()?)
     }
 
-    fn get_trace_from_current_thread(&self) -> Result<StackTrace, MemoryCopyError> {
+    #[cfg(any(target_os = "macos", target_os = "windows", target_os = "linux"))]
+    fn is_on_cpu_os_specific(&self) -> Result<bool> {
+        // remoteprocess crate exposes a Thread.active() method for each of these targets
+        for thread in self.process.threads()?.iter() {
```

Comment:
This is not the logic you want - you need to check the "current" thread that we're going to get stacktraces from, not just any thread in the process

Reply:
We don't know which OS thread the "current" ruby thread until we start examining the interpreter's memory, so the heuristic is: "if any thread in the thread group is on CPU"

Reply:
If the "current" ruby thread is not active OS-wise, it's surely not active ruby-wise (to be precise: it may be marked active ruby-wise, this just means it's preempted). Another thread may be running, doing non-ruby stuff, and the thread doing ruby-stuff may be preempted, and you will incorrectly account that as the ruby code)

Reply:
I touched on this in another comment, but if this is meant as an initial filter and the more accurate check happens later (via `get_thread_status`), then I think it's okay. Unless the remoteprocess functions we're calling here are expensive, this should reduce rbspy's CPU usage a bit when we're doing on-CPU profiling. It might be worth it to check that assumption, though -- I don't know offhand how expensive this loop is for a typical program.

Reply:
@acj this is just not accurate enough - imagine this case: 1 thread in the application is running some non-related C code doing whatever (and it's active, so `thread.active()` returns `true`). Meanwhile, another thread is running Ruby code (so it has `rb_thread_status_THREAD_RUNNABLE`, I guess?). But the OS decided that it's time that the Ruby thread relinquishes CPU, and it gets preempted. Its `rb_thread_status` remains as it was.

So, per the 2 checks:
* the "process" is active, because one of its threads is active.
* the "Ruby thread" is also "active" because its status is `THREAD_RUNNABLE`.

And the result is that we collect Ruby stacks from a thread that's not currently running!

About the cost of `thread.active()` - it's not very expensive. It's a single read from `/proc/tid/stat`, so 3 syscalls (`open`, `read`, `close`). Comparing to `get_thread_status` - which is only one `process_vm_readv` IIRC, so faster; but it should be benched & compared to reading `/proc/tid/stat` if we want a real comparison.

Reply:
Yeah, sorry, I didn't mean to suggest that the entire approach is okay. I understood your first example and agree that it's something we need to solve (I'm still thinking about this). What I meant to say is that this `for` loop seems reasonable, provided that we can somehow do an accurate check later when we examine process memory.

Reply:
> I touched on this in another comment, but if this is meant as an initial filter and the more accurate check happens later (via `get_thread_status`), then I think it's okay. Unless the remoteprocess functions we're calling here are expensive, this should reduce rbspy's CPU usage a bit when we're doing on-CPU profiling. It might be worth it to check that assumption, though -- I don't know offhand how expensive this loop is for a typical program.

Anyway, I used the [flame](https://crates.io/crates/flame) crate to insert tracepoints in `get_trace()` and this is what I had measured:
![image](https://user-images.githubusercontent.com/2903914/122680149-85177880-d1f6-11eb-9be8-acb01f4dd08f.png)
```
THREAD: 140546956667136
| get_trace: 0.246437ms
  | is_on_cpu_os_specific: 0.024342ms
  | get_stack_trace_from_current_thread: 0.221337ms
```

It seems like the remoteprocess functions a bit expansive to call per trace..

Reply:
@acj What is your take given this measurement ? should I get rid of the `remoteprocess` calls ?

Reply:
I don't know yet. On the one hand, the call to check active threads isn't free, but on the other hand we'd avoid the more expensive call to `get_stack_trace_from_current_thread` if the process is definitely not on-CPU. That might be a reasonable tradeoff for some workloads.

Let's revisit this after we solve the correctness/reliability questions.